janus/src/message.rs

```diff
@@ -226,6 +232,118 @@ impl Display for Interval {
+            Nonempty(Inclusive(None), _)
+            | Nonempty(Exclusive(None), _)
+            | Nonempty(_, Inclusive(None))
+            | Nonempty(_, Exclusive(None)) => {
+                Err("Interval cannot represent a timestamp range with a null bound".into())
+            }
+            Nonempty(Unbounded, _) | Nonempty(_, Unbounded) => {
+                Err("Interval cannot represent an unbounded timestamp range".into())
+            }
+            Nonempty(Exclusive(_), _) | Nonempty(_, Inclusive(_)) => {
+                Err("Interval can only represent timestamp ranges that are closed at the start and open at the end".into())
+            }
+            Nonempty(Inclusive(Some(start_raw)), Exclusive(Some(end_raw))) => {
+                // These timestamps represent the number of microseconds before (if negative) or
+                // after (if positive) midnight, 1/1/2000.
+                let start_timestamp = timestamp_from_sql(start_raw)?;
+                let end_timestamp = timestamp_from_sql(end_raw)?;
+
+                // Convert to Unix timestamp, in seconds since midnight 1/1/1970.
+                let negative = start_timestamp < 0;
+                let abs_start_us = start_timestamp.unsigned_abs();
+                let abs_start_secs = abs_start_us / USEC_PER_SEC;
+                let time = if negative {
+                    if abs_start_secs > TIME_SEC_CONVERSION {
+                        return Err("Interval cannot represent timestamp ranges starting before the Unix epoch".into());
+                    }
+                    Time::from_seconds_since_epoch(TIME_SEC_CONVERSION - abs_start_secs)
+                } else {
+                    Time::from_seconds_since_epoch(TIME_SEC_CONVERSION + abs_start_secs)
+                };
+
+                if end_timestamp < start_timestamp {
+                    return Err("timestamp range ends before it starts".into());
+                }
+                let duration_us = end_timestamp.abs_diff(start_timestamp);
+                let duration = Duration::from_seconds(duration_us / USEC_PER_SEC);
+
+                Ok(Interval::new(time, duration)?)
+            }
+        }
+    }
+
+    accepts!(TS_RANGE);
+}
+
+#[cfg(feature = "database")]
+impl ToSql for Interval {
+    fn to_sql(
+        &self,
+        _: &postgres_types::Type,
+        out: &mut bytes::BytesMut,
+    ) -> Result<postgres_types::IsNull, Box<dyn std::error::Error + Sync + Send>> {
+        let start_unix_timestamp_secs = i64::try_from(self.start().as_seconds_since_epoch())
+            .map_err(|_| "Interval start is out of range")?;
+        let duration_secs = i64::try_from(self.duration().as_seconds())
+            .map_err(|_| "Interval duration is out of range")?;
+
+        // Convert from the 2000 epoch to the 1970 epoch, and from seconds to microseconds.
+        let start_sql_usec = start_unix_timestamp_secs
```

Comment:
Similarly to above, if we had a const "zero SQL timestamp" `Time` value, we could just use `Time::difference` here to simplify the convert-epoch logic. A new `Duration::as_microseconds` would simplify the seconds-to-usec logic.